### PR TITLE
UploadV4 metadata workaround

### PR DESCRIPTION
**Pull request by ia**
* **Please check if the PR fulfills these requirements**
- [x] The changes have been tested locally
- [x] There are no breaking changes

* **What kind of change does this PR introduce?**
Fixing implemented by design error in upload V4.

* **What is the current behavior?**
[_*it is no longer possible to upload to the same named Artifact multiple times. You must either split the uploads into multiple Artifacts with different names, or only upload once. Otherwise you will encounter an error.*_](https://github.com/actions/upload-artifact/blob/main/README.md#breaking-changes)
Nice.

* **What is the new behavior (if this is a feature change)?**
Metadata is zip'ped in output artifacts for [each model from every matrix individually](https://github.com/ia/IronOSf/actions/runs/7255681052#artifacts).

* **Other information**:
~I tried to workaround this with _cache_ action but couldn't figure out how to make it work properly. Another option is to toss those JSONs like hot potatoes between jobs and steps but it will bloat `push.yml`. Regardless any solution, implementation of _job queue in line_ will be required anyway because before even trying to upload all of JSONs at once even when there is a way to toss all of them into one step, this exact step must be waiting all other builds to be finished first <<< and this will have to be implemented somehow as well just as managing to stop this waiting step if some other build step has failed. Hence, just a quick&dirty solution for now to highlight the issue.~

What is the purpose of JSON anyway? I think I saw somewhere here that it used to be used by some 3rd party tools/updaters. Is this still valid? Is _metadata.zip_ (as it is) really necessary? In case of this commit I guess we will find out soon anyway.

---

P.S. Permission to speak freely, capt'n? Just can't leave it without a rant.

Expectation (as seen on the internet): AI-based ML-driven Automagitical HighTech will ease development and make it as breeze with brand new tools & languages!

Reality: a meatbag digging a couple of hours in documentation & configs trying samples & running testing builds only because `v3` has been replaced by `v4`. :roll_eyes:

### CODE REVIEW

**Comment by ia** [comment]
[More information](https://github.com/actions/upload-artifact/issues/472) of this wonderful _not-a-bug_ feature. If I understand the discussion and [this "migration" official document](https://github.com/actions/download-artifact/blob/main/docs/MIGRATION.md) *correctly*, they did add ability to fetch multiple artifacts at once for another _job_ BUT not uploading of multiple files in one `artifact.zip` anyway. However, I will try to test it but I guess it won't work.

**Comment by ia** [comment]
Hmm... _it seems_ [it's working](https://github.com/ia/IronOSf/actions/runs/7256395299#artifacts). I guess bunch of `metadata-*` is not much of a problem in artifacts in comparison with preserving backward compatibility.

Oh, and one more thing, just in case if it will be merged: last time the build steps in `push.yml` have been changed, some settings inside of the whole project had to be tweaked as well for further build runs on PRs/merges correctly.

**Comment by Ralim** [comment]
Hia,
I'll have to look at this later on.
The `metadata.zip` file _cant_ be changed at this point in time as its locked in for software that isnt trivial to update.
(i.e it must contain all the json files).

I was thinking worst case might need to make a final step that pulls the previous artefacts to generate it :/

**Comment by ia** [comment]
> Hia, I'll have to look at this later on.

Sure, no problem! Take your time.

> The `metadata.zip` file _cant_ be changed at this point in time

With this PR it's generated as it used to be - just check `metadata` package [here](https://github.com/Ralim/IronOS/actions/runs/7256643720#artifacts) to make sure that everything as it must be.

> its locked in for software that isnt trivial to update.

Could you, please, elaborate more in details (when you have time), about what software are you talking about exactly? Just very curious now.

> I was thinking worst case might need to make a final step that pulls the previous artefacts to generate it :/

No-no. So, the way how _this_ solution works is that:
- in `build` & `build_multi-lang` steps every JSON uploaded separately (i.e., `metadata-TS100.zip` with `TS100.json`, `metadata-TS101.zip`, etc. <<< this is the only _tradeoff_ but that's what GitHub basically suggests as workaround)
- in new step - `upload_metadata`:
  - every separate but already uploaded JSON file downloaded again & unpacked
  - after all of them are downloaded, they are repacked & re-uploaded but at this time in `metadata.zip`
  - maybe this can be considered as the second _tradeoff_ but this build step will have to wait for `build` & `build_multi-lang` steps first obviously so no fancy multitasking like "build & upload everything at once".

Hope this helped.

**Comment by Ralim** [comment]
>Could you, please, elaborate more in details (when you have time), about what software are you talking about exactly? Just very curious now.

Factory tooling during manufacture of Pinecil's reads this zip file, and the older Pine64 flasher tool also uses it to know what language options to present users.


This solution is what I expect, sorry I missed the re-merge at the end.
I wonder if we can remove the other artefacts at the end to make the list cleaner for when I do releases 🤔

**Comment by ia** [comment]
> Factory tooling during manufacture of Pinecil's reads this zip file

During manufacture? **Right from this github project**? Wooow. How interesting!!! But it's just basically _lang name - file name_ key-value data file... why is this file sooo essential exactly, may I ask (if it's not some NDA secret, of course)?..
 
> sorry I missed the re-merge at the end.

No-no-no, **I am** sorry bombarding here with comments & rants straying away the discussion.

> I wonder if we can remove the other artefacts at the end to make the list cleaner for when I do releases 🤔

Yeaaa... I know, right? Hmmm... let me try one more thing then.

**Comment by ia** [comment]
You see, here is the problem with using `cache` through `save` / `restore`:
- [here](https://github.com/Ralim/IronOS/actions/runs/7257575500/job/19771607924#step:13:11) `save` tells that it's not only there but it already exists;
- but [here](https://github.com/Ralim/IronOS/actions/runs/7257575500/job/19771640980#step:6:9) `restore` tells that it doesn't.

I ran out of ideas how to make it work except only one for now...

**Comment by ia** [comment]
Unfortunately, from my side this is the best solution which I ended up with this issue, so:
- individual metadata artifacts stays;
- `metadata.zip` _seems_ to be valid.

However, now individual metadata artifacts _should be kept_ only for one day.

As probable other solution - reworking `push.yml`: merge `build` & `build_multi-lang` into one job, build everything in one "session" and then upload in bulk all needed files to artifacts with necessary names. But it doesn't look good neither, I guess. At least unnecessary artifacts can be deleted manually right through github interface.

**Comment by ia** [comment]
I think I have idea how to reduce _amount_ of intermediate individual metadata files in the artifacts, at least. Not sure if it will work though but will try later today.

**Comment by ia** [comment]
TL; DR: explanation in one image to whom who may concern & curious but didn't have time nor interest to track the latest GitHub :tm: Actions Platform changes:  




---

Bloody hell!.. pardon my French.

So... Machines: 0, Meatbags: 1

... but only if it's affordable to put one tiny `JSON` file per each artifact for every model... I hope :crossed_fingers: *it is* since there are already files like `LICENSE`, etc. _Is it affordable, **right**_? _(Princess Amidala Face meme)_

In the essence, with [**the latest commit here**](https://github.com/Ralim/IronOS/pull/1855/commits/ed3ac204ca9b669a0fb35ecc6c38e7f456db271c) the following actions take place:
- generate `JSON` for every model in the existed steps;
- upload `JSON` for every model from the existed steps altogether with other files (inside artifact per model like before)
- add new _job_ [`upload_metadata`](https://github.com/Ralim/IronOS/blob/ed3ac204ca9b669a0fb35ecc6c38e7f456db271c/.github/workflows/push.yml#L106), where:
  - wait for & download all the artifacts
  - pack & upload _`JSON` files only_ from already prebuilt artifacts as `metadata.zip`
  - HOWEVER, this _relies_ on the very important _assumption_ (i.e. **_current_** GitHub :tm:  Actions **behavior**): not setting `pattern` directive (which means which artifacts exactly should be downloaded for this step) means _download all the ~things~ artifacts_.

I did compare manually the files from `metadata.zip` [here (current `dev`)](https://github.com/Ralim/IronOS/actions/runs/7256912663) and [here (the latest commit as of writing)](https://github.com/Ralim/IronOS/actions/runs/7270237583) myself since **it's very essential**. But besides changes in the _tag_ (**how it should be**) these `metadata.zip`'s are identical. However, for some reason GitHub :tm:  shows for me different (_web page_ VS _actual_) size [here (current `dev`)](https://github.com/Ralim/IronOS/actions/runs/7256912663): the page says _**~40K**_ size for `metadata.zip` but when I download it, it's just like it should which is _about_ _**~8K**_. :man_shrugging: 

---

P.S. Just out of curiosity... is it really necessary to bump all the github "modules" (like _checkout_, _upload_, etc.) to the latest versions or no? Is my understanding correct that if you don't do that - eventually you just loose the ability to use the older versions at all, right?

**Comment by ia** [comment]
In short, how the proposed solution works (which is the latest commit at the time of writing this comment):
- put `model.json` to artifact with `model` (including multi-lang builds for Pinecils) which modifies `build` & `build_multi-lang` jobs)
- in addition step (`upload_metadata`):
  - wait for all builds;
  - download all the artifacts;
  - reapload JSON files from downloaded artifacts as `metadata.zip`.

**Comment by ia** [comment]
Hello. I know that probably everyone is busy now with holidays. But just out of curiosity I would like to know: will there be switch to _checkoutV4_ _soon_ or _checkoutV3_ will be kept so I should just close this PR? Thanks.

**Comment by Ralim** [comment]
Hia, I plan to merge this in now, can you drop the image out of the PR?

**Comment by ia** [comment]
> Hia, I plan to merge this in now

Aha, good. Are you sure that you're OK with the implemented solution? If you have a better idea, just let me know and we will figure out something.

> can you drop the image out of the PR?

Wait... what? Ooooooh! I feel so-so-so embarrassed right now! I thought I did revert this mess with [these](https://github.com/Ralim/IronOS/pull/1855/commits/f3978e0350ead328f45ac417973a8de94c3cc2ac) [commits](https://github.com/Ralim/IronOS/pull/1855/commits/2927780567e5c084ecc144a8445c42e79415202a). **Sure I can!** Sorry so much. Gosh... that's awful of me, obviously it doesn't belong there definitely. I guess I was getting too emotional when github team did break such a common use case for its users so I even didn't notice that the image ended up & still in the repo. Will try to avoid such behavior in the future.  That was totally not appropriate. :|

**Comment by Ralim** [comment]
>Wait... what? Ooooooh! I feel so-so-so embarrassed right now! I thought I did revert this mess with [these](https://github.com/Ralim/IronOS/pull/1855/commits/f3978e0350ead328f45ac417973a8de94c3cc2ac) [commits](https://github.com/Ralim/IronOS/pull/1855/commits/2927780567e5c084ecc144a8445c42e79415202a). Sure I can! Sorry so much. Gosh... that's awful of me, obviously it doesn't belong there definitely. I guess I was getting too emotional when github team did break such a common use case for its users so I even didn't notice that the image ended up & still in the repo. Will try to avoid such behavior in the future. That was totally not appropriate. :|

Honestly, I feel you; its infuriating when they break stuff like this without a good solution to fix it, so I dont mind the rant at all 🤣 



>Aha, good. Are you sure that you're OK with the implemented solution? If you have a better idea, just let me know and we will figure out something.

I dont think we will get a better solution _really_ for a while

**Review by Ralim — APPROVED** [review]
